### ISSUE TITLE
Try to get pyhf to run in Pyolite / JupyterLite

### ISSUE THREAD

**Issue by matthewfeickert** [scikit-hep/pyhf#1775]
### Summary

As [`pyodide` is now running with `scipy` `v1.7.3`](https://github.com/pyodide/pyodide/blob/950c1dcbd271f90faea0d205a7c95fcd40710a69/packages/scipy/meta.yaml#L1-L3) ([demoed on the NumPy docs](https://twitter.com/jtpio/status/1491698955859767298)) all the components to be able to use `pyhf` in WebAssembly should be there. It would be interesting to see how much work would be needed to get a minimal working example going and then to run some examples in the docs with [JupyterLite](https://github.com/jupyterlite/jupyterlite).

cc @kratsg as he was thinking the same thing as me when he saw the NumPy tweet.

### Code of Conduct

- [X] I agree to follow the Code of Conduct

**Comment by matthewfeickert** [comment]
@henryiii points out that you "can pure python packages with micropip" as [the docs mention](https://pyodide.org/en/stable/usage/packages-in-pyodide.html)

> Note that in addition to this list, pure Python packages with wheels can be loaded directly from PyPI with [micropip.install](https://pyodide.org/en/stable/usage/api/micropip-api.html#micropip.install).

**Comment by henryiii** [comment]
![Screen Shot 2022-02-11 at 12 11 20 PM](https://user-images.githubusercontent.com/4616906/153637039-745b8d2a-8d5c-4754-8ff6-9802ed4c0f16.png)

**Comment by henryiii** [comment]
Anything with plots that doesn't use data? scrapbook is not supported, due to pyzmq.

**Comment by matthewfeickert** [comment]
This is awesome!

```python
import micropip
await micropip.install("pyhf[contrib]")

import pyhf
from pyhf.contrib.viz import brazil
import numpy as np
import matplotlib.pyplot as plt

model = pyhf.simplemodels.uncorrelated_background(
    signal=[5.0, 10.0], bkg=[50.0, 60.0], bkg_uncertainty=[5.0, 12.0]
)
observations = [53.0, 65.0] + model.config.auxdata

poi_values = np.linspace(0.1, 5, 50)
obs_limit, exp_limits, (scan, results) = pyhf.infer.intervals.upperlimit(
    observations, model, poi_values, level=0.05, return_results=True
)
print(f"Upper limit (obs): μ = {obs_limit:.4f}")
print(f"Upper limit (exp): μ = {exp_limits[2]:.4f}")

fig, ax = plt.subplots()
fig.set_size_inches(10.5, 7)
ax.set_title("Hypothesis Tests")

artists = brazil.plot_results(poi_values, results, ax=ax)
plt.show()
```

![pyhf_webassm](https://user-images.githubusercontent.com/5142394/153640635-6f058acb-7ab9-4f2d-a103-83a166fee128.png)

**Comment by matthewfeickert** [comment]
As @lukasheinrich points out:

> This is amazing for HEPData integration. Imagine a "prefit/posfit" stack plot visualization off the likelihood displayed right on HEPData.

**Comment by kratsg** [comment]
Even better -- the example in the issue encoded in the URL: https://replite.vercel.app/retro/consoles/?toolbar=1&kernel=python&code=import%20micropip%0Aawait%20micropip.install%28%22pyhf%22%29%20%20%23%20pyodide%20has%20scipy%20and%20matplotlib%20already%0A%0Aimport%20pyhf%0Afrom%20pyhf.contrib.viz%20import%20brazil%0Aimport%20numpy%20as%20np%0Aimport%20matplotlib.pyplot%20as%20plt%0A%0Amodel%20%3D%20pyhf.simplemodels.uncorrelated_background%28%0A%20%20%20%20signal%3D%5B5.0%2C%2010.0%5D%2C%20bkg%3D%5B50.0%2C%2060.0%5D%2C%20bkg_uncertainty%3D%5B5.0%2C%2012.0%5D%0A%29%0Aobservations%20%3D%20%5B53.0%2C%2065.0%5D%20%2B%20model.config.auxdata%0A%0Apoi_values%20%3D%20np.linspace%280.1%2C%205%2C%2050%29%0Aobs_limit%2C%20exp_limits%2C%20%28scan%2C%20results%29%20%3D%20pyhf.infer.intervals.upperlimit%28%0A%20%20%20%20observations%2C%20model%2C%20poi_values%2C%20level%3D0.05%2C%20return_results%3DTrue%0A%29%0Aprint%28f%22Upper%20limit%20%28obs%29%3A%20%CE%BC%20%3D%20%7Bobs_limit%3A.4f%7D%22%29%0Aprint%28f%22Upper%20limit%20%28exp%29%3A%20%CE%BC%20%3D%20%7Bexp_limits%5B2%5D%3A.4f%7D%22%29%0A%0Afig%2C%20ax%20%3D%20plt.subplots%28%29%0Afig.set_size_inches%2810.5%2C%207%29%0Aax.set_title%28%22Hypothesis%20Tests%22%29%0A%0Aartists%20%3D%20brazil.plot_results%28poi_values%2C%20results%2C%20ax%3Dax%29%0Aplt.show%28%29

**Comment by matthewfeickert** [comment]
Thanks to @kratsg's tip on `urllib.parse` here's a bit of code to generate a URL that will [enable the `replite` toolbar](https://github.com/jtpio/replite#enable-the-toolbar) as well

```python
# pyolite_src.py
import urllib.parse

code = """\
import micropip
await micropip.install(["pyhf", "requests"])
import pyhf\
"""

parsed_url = urllib.parse.quote(code)
url_base = "https://replite.vercel.app/retro/consoles/?toolbar=1&kernel=python&code="

print(f"# replite URL:\n{url_base + parsed_url}")
```

[![lite-badge]][lite]

[lite-badge]: https://jupyterlite.rtfd.io/en/latest/_static/badge.svg
[lite]: https://replite.vercel.app/retro/consoles/?toolbar=1&kernel=python&code=import%20micropip%0Aawait%20micropip.install%28%5B%22pyhf%22%2C%20%22requests%22%5D%29%0Aimport%20pyhf

**Comment by matthewfeickert** [comment]
I'm trying to test thins on TestPyPI where we might be able to get HEPData to work until we have `v0.7.0` out but I'm hitting issues with installs, so I opened up this Issue: https://github.com/pyodide/pyodide/issues/2166

**Comment by matthewfeickert** [comment]
Looks like the https://replite.vercel.app/retro/consoles/?toolbar=1 URLs are now 404ing and we should be switching to using https://jupyterlite.readthedocs.io/en/latest/_static/repl/index.html?toolbar=1 as a base URL.

(I'll go and update the URLs above to keep the examples working.)

**Comment by matthewfeickert** [comment]
> We can stick this in an iframe perhaps

@kratsg was thinking ahead. :+1: 

Today I read Jeremy Tuloup's [Jupyter Everywhere](https://blog.jupyter.org/jupyter-everywhere-f8151c2cc6e8) blog post that he also put out today (2022-03-15) and with just 

```diff
$ git diff origin/master 
diff --git a/README.rst b/README.rst
index 07fe617c..763f0cc0 100644
--- a/README.rst
+++ b/README.rst
@@ -32,6 +32,17 @@ to support modern computational graph libraries such as PyTorch and
 TensorFlow in order to make use of features such as autodifferentiation
 and GPU acceleration.
 
+Try out now with JupyterLite
+----------------------------
+
+.. raw:: html
+
+   <iframe
+      src="https://jupyterlite.github.io/demo/repl/index.html?kernel=python&toolbar=1&code=import%20micropip%0Aawait%20micropip.install%28%5B%22pyhf%22%2C%20%22requests%22%5D%29%0Aimport%20pyhf"
+      width="100%"
+      height="500px"
+   ></iframe>
+
 User Guide
 ----------
 
```

things are running in the docs:

![in_your_docs_calculating_your_cls](https://user-images.githubusercontent.com/5142394/158461566-f5075d2f-7adc-4076-abb8-199193aa219a.png)

**Comment by kratsg** [comment]
> @kratsg was thinking ahead. 👍

not really. I just know the limitations of gh-pages and sphinx :)

**Comment by matthewfeickert** [comment]
The more interesting part of the article is about [`jupyterlite-sphinx`](https://github.com/jupyterlite/jupyterlite-sphinx). Though I haven't been able to get it working well yet.

**Comment by matthewfeickert** [comment]
I've also now come across why `requests` isn't working in `pyodide`: https://github.com/pyodide/pyodide/issues/529

So until that's fixed (probably won't be for quite some time as I think the `pyodide` team is already swamped and focusing on other areas currently) we'd need to acquire files from HEPData through an alternative method.

**Comment by matthewfeickert** [comment]
> So until that's fixed (probably won't be for quite some time as I think the `pyodide` team is already swamped and focusing on other areas currently) we'd need to acquire files from HEPData through an alternative method.

Actually this got addressed 4 days ago in https://github.com/pyodide/pyodide/pull/2263! So that's exciting. Hopefully the next(?) release of `pyodide` (`v0.19.2`) will include this.

**Comment by matthewfeickert** [comment]
@kratsg @lukasheinrich to follow up on some discussions that we've had in chat: I currently have something similar to https://github.com/scikit-hep/pyhf/issues/1775#issuecomment-1068415788 implemented on the branch `docs/embedd-jupyterlite-into-readme` that has the following RTD render (c.f. PR #1820)

https://pyhf.readthedocs.io/en/docs-embedd-jupyterlite-into-readme/

I tried using [jupyterlite-sphinx](https://github.com/jupyterlite/jupyterlite-sphinx), but it isn't ready yet given https://github.com/jupyterlite/jupyterlite-sphinx/issues/37 and so we'll have to wait for that later on down the line.

The current thing to figure out is the following:

As PR #1820 is currently using an iframe inside of a

```rst
.. raw:: html
```

directive in the README the iframe won't render on a GitHub preview render of the README, or on PyPI for security reasons. This means that if we have text around the iframe explaining it that that text will show up on GitHub and PyPI with no iframe and seem strange and out of context, or like our docs are broken. It would be good to avoid this.

The next question then is how to avoid this? Do we just include the directive with no supporting text and let it get ripped out/not rendered by GitHub and PyPI and assume that anyone visiting the docs that sees it will understand what it is without instructions? The NumPy team gives [quite a bit of helpful info in theirs](https://numpy.org/). Or do we stop using the `README.rst` as the landing page for our docs and then have a seperate page that is very similar to the README that actually gets rendered for the docs? This spares PyPI, but means that we now have two documents to keep synced. Either that or we need to then start doing injections of the raw HTML directive _into_ the README.rst at docs build time but that sounds like a messy idea.

Thoughts?

**Comment by kratsg** [comment]
Also link #1736 as it's related.

**Comment by kratsg** [comment]
> This spares PyPI, but means that we now have two documents to keep synced. Either that or we need to then start doing injections of the raw HTML directive _into_ the README.rst at docs build time but that sounds like a messy idea.

Why not just include it directly in `docs/index.rst` and not in the README? The README is what's shown for github and pypi, and it's injected into our docs. So we can add more in `index.rst` without screwing up the other places.

**Comment by kratsg** [comment]
> I could be missing something easy here though.

Splice it: https://stackoverflow.com/a/54519037/1532974

**Comment by matthewfeickert** [comment]
> Splice it: https://stackoverflow.com/a/54519037/1532974

:exploding_head:

**Comment by matthewfeickert** [comment]
> > So until that's fixed (probably won't be for quite some time as I think the `pyodide` team is already swamped and focusing on other areas currently) we'd need to acquire files from HEPData through an alternative method.
> 
> Actually this got addressed 4 days ago in [pyodide/pyodide#2263](https://github.com/pyodide/pyodide/pull/2263)! So that's exciting. Hopefully the next(?) release of `pyodide` (`v0.20.0` in April [pyodide/pyodide#529 (comment)](https://github.com/pyodide/pyodide/issues/529#issuecomment-1068563138)) will include this.

@henryiii [has pointed out](https://gitter.im/Scikit-HEP/community?at=623ddb2de9cb3c52ae325cfe) that this actually won't happen:

> No it won’t. Requests doesn’t work unless someone rewrites the stdlib module in terms of browser APIs.

and also pointed us to https://github.com/pyodide/pyodide/issues/140.

So looks like I got overly excited and didn't read https://github.com/pyodide/pyodide/issues/529#issuecomment-962175588 carefully enough.

**Comment by henryiii** [comment]
And you can make it work (not using portable code though):

```python
import pyodide
import json
inp = pyodide.open_url("https://raw.githubusercontent.com/scikit-hep/pyhf/master/docs/examples/json/2-bin_1-channel.json”)
wspace = pyhf.Workspace(json.load(inp))
```

(redirects are not supported)
